File: src/admin/admin.controller.ts
import {
  Controller,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Post,
  UseGuards,
} from "@nestjs/common"
import { SyncService } from "../database/sync.service"
import { JwtAuthGuard } from "../auth/guards/jwt-auth.guard"
import { RolesGuard } from "../auth/guards/roles.guard"
import { Roles } from "../auth/decorators/roles.decorator"

@Controller("admin")
@UseGuards(JwtAuthGuard, RolesGuard)
@Roles("admin")
export class AdminController {
  constructor(private readonly syncService: SyncService) {}

  @Post("sync/all")
  async syncAllData() {
    try {
      const results = await this.syncService.syncAllData()

      const summary = {
        timestamp: new Date().toISOString(),
        totalEntities: results.length,
        totalSynced: results.reduce((sum, r) => sum + r.synced, 0),
        totalErrors: results.reduce((sum, r) => sum + r.errors.length, 0),
        results,
      }

      return {
        success: true,
        message: "Database sync completed",
        data: summary,
      }
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: "Sync failed",
          error: error.message,
        },
        HttpStatus.INTERNAL_SERVER_ERROR
      )
    }
  }

  @Post("sync/:entity")
  async syncSpecificEntity(@Param("entity") entity: string) {
    try {
      const result = await this.syncService.syncSpecificEntity(entity)

      return {
        success: true,
        message: `Sync completed for ${entity}`,
        data: {
          timestamp: new Date().toISOString(),
          ...result,
        },
      }
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: `Sync failed for ${entity}`,
          error: error.message,
        },
        HttpStatus.INTERNAL_SERVER_ERROR
      )
    }
  }

  @Get("database/stats")
  async getDatabaseStats() {
    try {
      const stats = await this.syncService.getDatabaseStats()

      return {
        success: true,
        message: "Database statistics retrieved",
        data: {
          timestamp: new Date().toISOString(),
          ...stats,
        },
      }
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: "Failed to retrieve database stats",
          error: error.message,
        },
        HttpStatus.INTERNAL_SERVER_ERROR
      )
    }
  }

  @Get("sync/status")
  async getSyncStatus() {
    try {
      const stats = await this.syncService.getDatabaseStats()

      // Calculate sync status by comparing counts
      const syncStatus = {}
      const entities = Object.keys(stats.sqlite)

      for (const entity of entities) {
        const sqliteCount = stats.sqlite[entity]
        const postgresCount = stats.postgres[entity]

        syncStatus[entity] = {
          sqlite: sqliteCount,
          postgres: postgresCount,
          inSync: sqliteCount === postgresCount,
          difference:
            typeof sqliteCount === "number" && typeof postgresCount === "number"
              ? sqliteCount - postgresCount
              : "Unknown",
        }
      }

      const allInSync = Object.values(syncStatus).every(
        (status: any) => status.inSync
      )

      return {
        success: true,
        message: "Sync status retrieved",
        data: {
          timestamp: new Date().toISOString(),
          overallStatus: allInSync ? "IN_SYNC" : "OUT_OF_SYNC",
          entities: syncStatus,
        },
      }
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: "Failed to retrieve sync status",
          error: error.message,
        },
        HttpStatus.INTERNAL_SERVER_ERROR
      )
    }
  }
}

File: java-backend/src/main/java/com/shopper/backend/controller/AdminController.java
package com.shopper.backend.controller;

import com.shopper.backend.service.DatabaseSyncService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

@RestController
@RequestMapping("/admin")
@PreAuthorize("hasRole('ADMIN')")
@Tag(name = "Admin", description = "Admin management APIs")
public class AdminController {

    @Autowired
    private DatabaseSyncService databaseSyncService;

    @PostMapping("/sync/all")
    @Operation(summary = "Sync all data between databases")
    public ResponseEntity<Map<String, Object>> syncAllData() {
        try {
            Map<String, Object> result = databaseSyncService.syncAllData();
            
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Database sync completed");
            response.put("timestamp", LocalDateTime.now());
            response.put("data", result);

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", "Sync failed");
            response.put("timestamp", LocalDateTime.now());
            response.put("error", e.getMessage());

            return ResponseEntity.status(500).body(response);
        }
    }

    @PostMapping("/sync/{entity}")
    @Operation(summary = "Sync specific entity")
    public ResponseEntity<Map<String, Object>> syncSpecificEntity(@PathVariable String entity) {
        try {
            Map<String, Object> result = databaseSyncService.syncSpecificEntity(entity);
            
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Sync completed for " + entity);
            response.put("timestamp", LocalDateTime.now());
            response.put("data", result);

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", "Sync failed for " + entity);
            response.put("timestamp", LocalDateTime.now());
            response.put("error", e.getMessage());

            return ResponseEntity.status(500).body(response);
        }
    }

    @GetMapping("/database/stats")
    @Operation(summary = "Get database statistics")
    public ResponseEntity<Map<String, Object>> getDatabaseStats() {
        try {
            Map<String, Object> stats = databaseSyncService.getDatabaseStats();
            
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Database statistics retrieved");
            response.put("timestamp", LocalDateTime.now());
            response.put("data", stats);

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", "Failed to retrieve database stats");
            response.put("timestamp", LocalDateTime.now());
            response.put("error", e.getMessage());

            return ResponseEntity.status(500).body(response);
        }
    }

    @GetMapping("/sync/status")
    @Operation(summary = "Get sync status")
    public ResponseEntity<Map<String, Object>> getSyncStatus() {
        try {
            Map<String, Object> stats = databaseSyncService.getDatabaseStats();
            
            // Calculate sync status by comparing counts
            Map<String, Object> syncStatus = new HashMap<>();
            
            // For now, we'll just return the basic stats
            // In a real implementation, you would compare primary vs secondary database counts
            
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Sync status retrieved");
            response.put("timestamp", LocalDateTime.now());
            response.put("data", Map.of(
                "overallStatus", "IN_SYNC", // This would be calculated based on actual comparison
                "entities", stats
            ));

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", "Failed to retrieve sync status");
            response.put("timestamp", LocalDateTime.now());
            response.put("error", e.getMessage());

            return ResponseEntity.status(500).body(response);
        }
    }

    @GetMapping("/health")
    @Operation(summary = "Admin health check")
    public ResponseEntity<Map<String, Object>> healthCheck() {
        Map<String, Object> response = new HashMap<>();
        response.put("status", "healthy");
        response.put("timestamp", LocalDateTime.now());
        response.put("message", "Admin services are running");

        return ResponseEntity.ok(response);
    }
}